### PR TITLE
Check validity of embed URLs

### PR DESCRIPTION
**Pull request by msarit**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This PR introduces improved error messaging when embed URLs fail to match supported URL types.
This PR fulfills [Task 1](https://github.com/forem/forem/issues/16687) of the four-task [Part 2](https://github.com/orgs/forem/projects/49/views/1) of the Unified Embeds epic.

## Related Tickets & Documents
- Closes https://github.com/forem/forem/issues/16687

## QA Instructions, Screenshots, Recordings
_[List of supported URLs](https://dev.to/p/editor_guide#liquidtags)_
Create an embed using:
(1) a supported URL with a typo. You should get an understandable error message related to the supported URL type.
(2) an unsupported URL without a typo. You should get the error: `Embeds for this URL are not supported`

NOTE: Eventually, the valid-but-unsupported-URL path will generate a fallback embed using OpenGraph metadata ([task 2](https://github.com/forem/forem/issues/16688)).

### UI accessibility concerns?
None

## Added/updated tests?

- [X] Not yet

## [Forem core team only] How will this change be communicated?
- [X] I will share this change internally with the appropriate teams

### CODE REVIEW

**Comment by msarit** [comment]
😩  😩  😩   So I used a mock GitHub account `aritmock` for some tutorials I created. apparently, I'm still authenticated using that account. Let me try and switch to my regular `msarit` account.

File: app/liquid_tags/unified_embed/tag.rb
Comment (dscottS3) [inline app/liquid_tags/unified_embed/tag.rb:54]: we may want to consider checking for `Net::HTTPNotFound` in addition to this. Only because I'm not entirely sure how it will handle a redirect. Essentially, if a provided url gets a 301 will it follow that redirect THEN return with 200? If so, we don't need to worry about it. If not, we'll have to add in maybe a couple more explicit checks here.

File: app/models/article.rb
Comment (dscottS3) [inline app/models/article.rb:585]: I'm not sure of the implications of this or if `Post Content` is the correct verbiage to use here. But for this case it sure looks a lot better than `base:` 😅

Comment (msarit) [reply app/models/article.rb:585]: Well seeing as it’s communicating errors, perhaps “Error” is better? 😄

File: app/liquid_tags/unified_embed/tag.rb
Comment (msarit) [reply app/liquid_tags/unified_embed/tag.rb:54]: Perhaps we can test a redirect response to see?

File: spec/support/embeds_helpers.rb
Comment (jgaskins) [inline spec/support/embeds_helpers.rb:10]: I like that you kept this simple instead of trying to anticipate all possible ways this could be customized in the future. I probably would've fallen into that trap. 😄

File: app/liquid_tags/loom_tag.rb
Comment (jgaskins) [inline app/liquid_tags/loom_tag.rb:23]: I like this a lot. Cleaning up the inputs when feasible is often better than raising a pedantic error. `git` does the pedantic thing, even when it knows what you want:

```
➜  forem git:(msarit/better-error-msg-invalid-embed-links) git di main... app -w
fatal: option '-w' must come before non-option arguments
```

The way you're doing it provides a better experience. 💯

File: app/models/article.rb
Comment (jgaskins) [reply app/models/article.rb]: Oh, I missed this part of the conversation because it was resolved. My [review comment](https://github.com/forem/forem/pull/16704#discussion_r820952069) asked about this change.

File: spec/liquid_tags/unified_embed/tag_spec.rb
Comment (msarit) [reply spec/liquid_tags/unified_embed/tag_spec.rb:4]: I will refactor! 🙏🏾

File: app/models/article.rb
Comment (msarit) [reply app/models/article.rb:585]: @jgaskins Ahhh thank you for this background information. When the error fires, it shows up as:
Screenshot from 2022-03-07 12-58-03
 You see the `base:` before the actual error message? So we felt it wasn't that descriptive as part of the error message.
What do you recommend we do then? I'm fine with undoing this change.

Comment (jgaskins) [reply app/models/article.rb:585]: @msarit Fascinating, I'm not sure. Do we need to show the attribute in the error message? Could we special-case `base`? I don't think `base` was ever intended to be displayed. Is that the Preact code rendering that or Rails rendering HTML?

File: app/javascript/article-form/components/ErrorList.jsx
Comment (msarit) [inline app/javascript/article-form/components/ErrorList.jsx:13]: In response to the concerns that @jgaskins raised [here](https://github.com/forem/forem/pull/16704#discussion_r820952069), I decided to render the `base` word conditionally in the error messaging, instead of changing `:base` to `Errors` like I had done before.

I'm asking our frontend folks @aitchiss @nickytonline to weigh in on this change. The change I made is very localized, affecting only the error messaging on Post Edits, and only for error messages of the attribute `:base`. However, if there is a reason for rendering the `base` word that I am missing, please inform me 🙏🏾 

Screen Shot 2022-03-08 at 10 46 19 AM

Comment (nickytonline) [reply app/javascript/article-form/components/ErrorList.jsx:13]: It makes sense not to do it on the backend as per [Jamie's comment](https://github.com/forem/forem/pull/16704#discussion_r820952069). Is this the only place we need to do this? If not consider a utility function, e.g. `sanitizeError`.

File: app/models/article.rb
Comment (jeremyf) [reply app/models/article.rb]: I think we want to keep `:base` around.  In ActiveRecord you can have an error on attributes or assigned to the `:base` object.  

https://guides.rubyonrails.org/active_record_validations.html#errors-base. Which strikes me that something is looking at the hash keys for rendering error messages.

File: app/javascript/article-form/components/ErrorList.jsx
Comment (jeremyf) [reply app/javascript/article-form/components/ErrorList.jsx:13]: See above comment, but the reason for base ties to how Rails looks at errors (either on an attribute or on the "base" record) See https://guides.rubyonrails.org/active_record_validations.html#errors-base

Comment (jeremyf) [reply app/javascript/article-form/components/ErrorList.jsx:13]: I like this approach and it preserves the idiomatic error key.

**Comment by msarit** [comment]
@jgaskins @jeremyf Could I get a last review/approval of this PR? 🙏🏾

**Review by jgaskins — APPROVED** [review]
Looks great! 💯

File: app/javascript/article-form/components/ErrorList.jsx
Comment (aitchiss) [reply app/javascript/article-form/components/ErrorList.jsx:13]: (non blocking) We could make this a bit more succinct, e.g. 

```jsx
<ul className="list-disc pl-6">
  {Object.keys(errors).map((key) => {
    return (
      <li key={key}>
        {key === 'base' ? errors[key] : `${key}: ${errors[key]}`}
      </li>
    );
  })}
</ul>
```

File: config/locales/liquid_tags/en.yml
Comment (aitchiss) [inline config/locales/liquid_tags/en.yml:102]: In testing I wasn't able to trigger this one 😄  I always got `not_found` - not sure what I'm missing in the test steps

**Review by aitchiss — APPROVED** [review]
Left a suggestion, and I had trouble triggering one of the errors, but maybe it's just me 😄 In any case, the errors I got back made sense in the context

File: config/locales/liquid_tags/en.yml
Comment (dscottS3) [reply config/locales/liquid_tags/en.yml:102]: This error should get triggered when a url doesn't exist. For example if one was to do `{% embed https://thisdomainreallydoesnotexist.wtf %} it should trigger that error.

Comment (aitchiss) [reply config/locales/liquid_tags/en.yml:102]: Can confirm that works 🙌  

I tried typoing some URLs that are normally valid (e.g. changing a single letter in a loom or github embed link), in which case I get something less readable, but I'm guessing this is just because it's genuinely triggering some more detailed/specific error - e.g.

error: SSL_connect SYSCALL returned=5 errno=0 state=SSLv3/TLS write client hello

Comment (msarit) [reply config/locales/liquid_tags/en.yml:102]: Ok so @aitchiss I was not able to reproduce your experience of that wonky error message. 🤷🏾‍♀️ 

I recorded a Loom of myself triggering the "URL not found" and "URL may have typo" error messages, and things seem to be working okay:

https://www.loom.com/share/a19a44700b764d438eda13d5760b69e7

I'll wait for feedback from @dscottS3 @aitchiss before proceeding to merge this PR 👍🏾

**Pull request merged**
